subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy

```diff
@@ -0,0 +1,248 @@
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'org.test:module:1.0' {
+                expectGetMetadata()
+                if (fixApplied) {
+                    expectGetArtifact()
+                }
+            }
+        }
+
+        then:
+        if (fixApplied) {
+            run ':checkDeps'
+            resolve.expectGraph {
+                root(":", ":test:") {
+                    module('org.test:module:1.0')
+                }
+            }
+        } else {
+            fails ':checkDeps'
+            failure.assertHasCause("Unable to find a matching configuration of org.test:module:1.0:")
+            failure.assertThatCause(containsNormalizedString("Required quality 'qa' and found incompatible value 'canary'"))
+        }
+
+        where:
+        fixApplied << [false, true]
+    }
+
+    @Unroll
+    def "variant attributes take precedence over component attributes (component level = #componentLevel)"() {
+        given:
+        repository {
+            'org.test:module:1.0'()
+        }
+        buildFile << """
+            def usage = Attribute.of('org.gradle.usage', String)
+            dependencies {
+                components {
+                    withModule('org.test:module') {                       
+                        if ($componentLevel) {
+                            withAttributes { attribute usage, 'unknown' }
+                        } else {
+                            withVariant('api') { withAttributes { attribute usage, 'unknownApiVariant' } }
+                            withVariant('runtime') { withAttributes { attribute usage, 'unknownRuntimeVariant' } }
+                        }
+                    }
+                }
+                conf 'org.test:module:1.0'
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'org.test:module:1.0' {
+                expectGetMetadata()
+                if (componentLevel) {
+                    expectGetArtifact()
+                }
+            }
+        }
+
+        then:
+        if (componentLevel) {
+            run ':checkDeps'
+            resolve.expectGraph {
+                root(":", ":test:") {
+                    module('org.test:module:1.0')
+                }
+            }
+        } else {
+            fails ':checkDeps'
+            failure.assertHasCause("Cannot choose between the following configurations of org.test:module:1.0:")
+            failure.assertThatCause(containsNormalizedString("Found org.gradle.usage 'unknownApiVariant' but wasn't required"))
+            failure.assertThatCause(containsNormalizedString("Found org.gradle.usage 'unknownRuntimeVariant' but wasn't required"))
+        }
+
+        where:
+        componentLevel << [true, false]
+    }
+
+    // This is eventually what we want, but version listing is not wired properly for this
+    @Ignore
+    def "can use a component metadata rule to infer quality attribute"() {
```

Comment:
:x: Should add test coverage for consuming a component attribute via `ComponentSelectionRules`.